### PR TITLE
[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #84423 [8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations
* #83916 [7/N] [Dispatchable Collectives] Update reduce with CPU / CUDA implementations
* #83876 [6/N] [Dispatchable Collectives] Update recv with CPU / CUDA implementations
* #83859 [5/N] [Dispatchable Collectives] Update send with CPU / CUDA implementations
* **#83810 [4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations**
* #83735 [3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations

### About this PR
* Update the all_reduce op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
* Update test to validate that a separate device implementation is not supported.

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D39506979](https://our.internmc.facebook.com/intern/diff/D39506979)

### CODE REVIEW

File: torch/csrc/distributed/c10d/OpsImpl.cpp [hunk 1]
```diff
@@ -42,6 +68,24 @@ TORCH_LIBRARY_IMPL(c10d, CPU, m) {
...
+
+TORCH_LIBRARY_IMPL(c10d, CPU, m) {
+  m.impl("allreduce_", allreduce_cpu_);
+}
+
+// TODO: The SparseCPU/SparseCUDA dispatched methods are only used to support
+// sparse all_reduce in the Gloo backend
+TORCH_LIBRARY_IMPL(c10d, SparseCPU, m) {
+  m.impl("allreduce_", allreduce_cpu_);
+}
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/OpsImpl.cpp:80]: nit: is there actually a `SparseCPU` or `SparseCUDA` implementation in Gloo?
If not, do we need to add it here?

**Review by kwen2501 — APPROVED** [review]
LGTM!

File: torch/csrc/distributed/c10d/OpsImpl.cpp (hunk 1 above)
Comment (H-Huang) [reply torch/csrc/distributed/c10d/OpsImpl.cpp:80]: There is logic within `allreduce()` to handle sparse tensors (https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupGloo.cpp#L1465-L1467) to which it branches to different logic, so having each implementation just call allreduce_ keeps behavior the same, but I think there may be a cleaner way to do this, we should discuss this.

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39506979).

File: torch/csrc/distributed/c10d/OpsImpl.cpp [hunk 2]
```diff
@@ -49,6 +84,24 @@ TORCH_LIBRARY_IMPL(c10d, CPU, m) {
...
+
+TORCH_LIBRARY_IMPL(c10d, SparseCUDA, m) {
```
Comment (kumpera) [inline torch/csrc/distributed/c10d/OpsImpl.cpp:98]: Why do we have allreduce_cuda_ and allreduce_cpu_ when they look identical?

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39506979).

File: torch/csrc/distributed/c10d/OpsImpl.cpp (hunk 2 above)
Comment (H-Huang) [reply torch/csrc/distributed/c10d/OpsImpl.cpp:98]: Right now is setting up the skeleton to have dispatching route to different implementations for CPU/CUDA. Yes the implementations are identical for now, they will be updated in the future to perform `process_group->backend[<cpu>/<cuda>]->all_reduce`, but right now it is a no-op.

**Pull request closed without merging**